### ISSUE TITLE
Edge-case race condition in WebProxy

### ISSUE THREAD

**Issue by MihaZupan** [dotnet/runtime#75134]
#73803 (7.0) introduced a race condition when recreating the internal `Regex` list.
The issue is that we update the `IsChanged` flag before we replace the `_regexBypassList` field [here](https://github.com/dotnet/runtime/blob/4da3eae7a54a4a6abeb59df4953b6de4c65b53ea/src/libraries/System.Net.WebProxy/src/System/Net/WebProxy.cs#L128-L145).
If two threads query `GetProxy` or `IsBypassed` at the same time, one may not observe changes to the bypass list and return the wrong answer.

```c#
var proxy = new WebProxy("myProxy:1234");
proxy.BypassArrayList.Add("contoso");

var task1 = Task.Run(() => proxy.IsBypassed(new Uri("http://contoso")));
var task2 = Task.Run(() => proxy.IsBypassed(new Uri("http://contoso")));

if (await task1 != await task2)
{
    Console.WriteLine("Oh no");
}
```

This is limited to cases where the user modified the bypass list via the `BypassArrayList` property, which is hopefully rare.

**Comment by stephentoub** [comment]
Accessing a WebProxy instance while its bypass list is being changed concurrently is already not thread safe and isn't supported.

**Comment by MihaZupan** [comment]
> is the issue here more that because it's lazily initialized this can happen even without concurrent user modification of the list?

Yes.

```c#
var proxy = new WebProxy("myProxy:1234");
proxy.BypassArrayList.Add("contoso");

myHandler.Proxy = proxy;
// ... errors can occur when making requests
```

**Comment by stephentoub** [comment]
I see; yeah, that's worth fixing.

**Comment by karelz** [comment]
Triage: Rare reliability problem. However, it is regression against 6.0. We should try to fix it in 7.0 if the fix is low-risk.

**Comment by liveans** [comment]
Reopening for backport to 7.0

**Comment by liveans** [comment]
Merged into release/7.0.
Closing

**Comment by karelz** [comment]
Fixed in main (8.0) in PR #75193 and in 7.0 (RC2) in PR #75246.
